test/AST-Quake/ranged_for.cpp

```diff
@@ -0,0 +1,210 @@
+    double sum = 0.0;
+
+    for (auto &d : parms) {
+      sum = sum + d;
+    }
+
+    // The above loop is syntactic sugar for the following loop.
+    //
+    // for (std::size_t i = 0; i < parms.size(); i++) {
+    //   const double &d = parms[i];
+    //   sum += d;
+    // }
+
+    rx(sum, q);
+    mz(q);
+  }
+};
+
+// CHECK-LABEL:   func.func @__nvqpp__mlirgen__Lt_Colonel(
+// CHECK-SAME:      %[[VAL_0:.*]]: !cc.stdvec<f64>) attributes
+// CHECK-DAG:       %[[VAL_1:.*]] = arith.constant 1 : index
+// CHECK-DAG:       %[[VAL_2:.*]] = arith.constant 0 : index
+// CHECK-DAG:       %[[VAL_3:.*]] = arith.constant 0.000000e+00 : f64
+// CHECK-DAG:       %[[VAL_4:.*]] = quake.alloca !quake.ref
+// CHECK-DAG:       %[[VAL_5:.*]] = cc.alloca f64
+// CHECK:           cc.store %[[VAL_3]], %[[VAL_5]] : !cc.ptr<f64>
+// CHECK:           %[[VAL_6:.*]] = cc.stdvec_size %[[VAL_0]] : (!cc.stdvec<f64>) -> i64
+// CHECK:           %[[VAL_7:.*]] = cc.stdvec_data %[[VAL_0]] : (!cc.stdvec<f64>) -> !cc.ptr<!cc.array<f64 x ?>>
+// CHECK:           %[[VAL_8:.*]] = arith.index_cast %[[VAL_6]] : i64 to index
+// CHECK:           %[[VAL_9:.*]] = cc.loop while ((%[[VAL_10:.*]] = %[[VAL_2]]) -> (index)) {
+// CHECK:             %[[VAL_11:.*]] = arith.cmpi slt, %[[VAL_10]], %[[VAL_8]] : index
+// CHECK:             cc.condition %[[VAL_11]](%[[VAL_10]] : index)
+// CHECK:           } do {
+// CHECK:           ^bb0(%[[VAL_12:.*]]: index):
+// CHECK:             %[[VAL_13:.*]] = arith.index_cast %[[VAL_12]] : index to i64
+// CHECK:             %[[VAL_14:.*]] = cc.compute_ptr %[[VAL_7]]{{\[}}%[[VAL_13]]] : (!cc.ptr<!cc.array<f64 x ?>>, i64) -> !cc.ptr<f64>
+// CHECK-DAG:         %[[VAL_15:.*]] = cc.load %[[VAL_5]] : !cc.ptr<f64>
+// CHECK-DAG:         %[[VAL_16:.*]] = cc.load %[[VAL_14]] : !cc.ptr<f64>
+// CHECK:             %[[VAL_17:.*]] = arith.addf %[[VAL_15]], %[[VAL_16]] : f64
+// CHECK:             cc.store %[[VAL_17]], %[[VAL_5]] : !cc.ptr<f64>
+// CHECK:             cc.continue %[[VAL_12]] : index
+// CHECK:           } step {
+// CHECK:           ^bb0(%[[VAL_18:.*]]: index):
+// CHECK:             %[[VAL_19:.*]] = arith.addi %[[VAL_18]], %[[VAL_1]] : index
+// CHECK:             cc.continue %[[VAL_19]] : index
+// CHECK:           } {invariant}
+// CHECK:           %[[VAL_20:.*]] = cc.load %[[VAL_5]] : !cc.ptr<f64>
+// CHECK:           quake.rx (%[[VAL_20]]) %[[VAL_4]] : (f64, !quake.ref) -> ()
+// CHECK:           %[[VAL_21:.*]] = quake.mz %[[VAL_4]] : (!quake.ref) -> i1
+
+struct Qernel {
+  void operator()(std::vector<double> parms) __qpu__ {
+    cudaq::qubit q;
+    double sum = 0.0;
+
+    for (auto &d : parms) {
+      d = sqrt(d);
+    }
+
+    // The above loop is syntactic sugar for the following loop.
+    //
+    // for (std::size_t i = 0; i < parms.size(); i++) {
+    //   double &d = parms[i];
+    //   d = sqrt(d);  // assignment to values held by parms
+    // }
+
+    for (auto d : parms) {
+      sum = sum + d;
+      d += 1000.0;
+    }
```

Comment:
This may be a dumb question, but will nested for loops work? I.e.
```cpp
for (auto d : params)
  for (auto j : params)
    sum += d*j;
```
Part of the reason I ask is because of the `[&]` lambda functions used for the implementation. (I think the answer is that they would work just fine ... I'm just wanting to double check.)

Reply:
Good question. They should work just fine, but we should also add a regression test. I'll add that to the issue.